core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertFalse(blocked.isDone());
+
+            buffer.taskFailed(taskId, new RuntimeException());
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked1 = buffer.isBlocked();
+            ListenableFuture<Void> blocked2 = buffer.isBlocked();
+            assertFalse(blocked1.isDone());
+            assertFalse(blocked2.isDone());
+
+            blocked2.cancel(true);
+
+            assertFalse(blocked1.isDone());
+            assertTrue(blocked2.isDone());
+        }
+    }
+
+    @Test
+    public void testPollPage()
+    {
+        testPollPages(ImmutableListMultimap.of(), ImmutableMap.of(), ImmutableList.of());
+        testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0")),
+                ImmutableMap.of(),
+                ImmutableList.of("p0a0v0"));
+        testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(),
+                ImmutableList.of("p0a1v0"));
+        testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(1, 0),
+                        createPage("p1a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(),
+                ImmutableList.of("p0a1v0"));
+        testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(1, 0),
+                        createPage("p1a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(
+                        createTaskId(2, 0),
+                        new RuntimeException("error")),
+                ImmutableList.of("p0a1v0"));
+        RuntimeException error = new RuntimeException("error");
+        assertThatThrownBy(() -> testPollPages(
```

Comment:
You should have two methods for testing positive and negative flow:
* testPollPages
* testPollPagesFailure

First one would take `List<String> expectedValues`
And the second one `Throwable expectedFailure`